### PR TITLE
Feature/#35 happy path for posts

### LINKED ISSUE #35: create posts component(s) for UI - API happy path
Create components to handle crud operations for posts and interact with the revconnect API server. 
GET, DELETE, PUT, POST

### PR DESCRIPTION
**Pull request by themellowpea**
### **TL;DR:**
- **Implemented the happy path for the posts endpoint**: All CRUD operations (Create, Read, Update, Delete) are now connected to the API.
- **Optimized state management**: Leveraged `useMemo` in the `PostsContext` to ensure efficient re-renders and better performance.
- **Refactored components**: Integrated context and hooks for managing posts, making the components more modular and scalable.

### **Description:**

This PR introduces the core functionality for managing posts through the API, focusing on establishing the `/posts` endpoint's happy path. The key changes include:

1. **Connected Posts CRUD Operations to API**:
   - **CreatePost**: Allows users to create new posts. Successfully created posts are now sent to the API, and the local state is updated accordingly.
   - **Read (FetchPosts)**: Posts are fetched from the API and displayed to the user. Pagination has been implemented, allowing users to navigate through different pages of posts.
   - **UpdatePost**: Users can edit existing posts. Changes are sent to the API, and the updated post is reflected in the UI.
   - **DeletePost**: Users can delete posts, and the changes are propagated to the API, with the local state updated to reflect the deletion.

2. **State Management Optimizations**:
   - Implemented `useMemo` in the `PostsContext` to optimize performance. This ensures that context values are only recalculated when necessary, reducing unnecessary re-renders across the application.

3. **Refactoring and Modularization**:
   - Refactored components like `CreatePost`, `PostPage`, and others to utilize hooks and context more effectively. This improves code readability and maintainability.
   - Encapsulated logic for handling posts in custom hooks (`useCreatePost`, `useUpdatePost`, `useDeletePost`, `useFetchPosts`), ensuring that components remain focused on rendering and UI logic.

### **Examples:**

1. **CreatePost Component Example**:

   The `CreatePost` component now utilizes the `PostsContext` to handle post creation:

   ```typescript
   import React, { useState } from "react";
   import { usePostsContext } from "../../hooks/usePostsContext";

   export default function CreatePost() {
     const [newTitle, setNewTitle] = useState("");
     const [newContent, setNewContent] = useState("");
     const { createPost } = usePostsContext();

     async function handleCreate(e: React.FormEvent) {
       e.preventDefault();
       const newPost = { title: newTitle, content: newContent };
       await createPost(newPost);
     }

     return (
       <form onSubmit={handleCreate}>
         <input
           type="text"
           value={newTitle}
           onChange={(e) => setNewTitle(e.target.value)}
           placeholder="Title"
           required
         />
         <textarea
           value={newContent}
           onChange={(e) => setNewContent(e.target.value)}
           placeholder="Content"
           required
         />
         <button type="submit">Create Post</button>
       </form>
     );
   }
   ```

2. **PostsContext with `useMemo`**:

   The `PostsContext` has been updated to use `useMemo`, optimizing the performance of context value updates:

   ```typescript
   import { createContext, ReactNode, useMemo, useState } from "react";
   import { useFetchPosts } from "../hooks/useFetchPosts";
   import { useCreatePost } from "../hooks/useCreatePost";
   import { useUpdatePost } from "../hooks/useUpdatePost";
   import { useDeletePost } from "../hooks/useDeletePost";

   export const PostsContext = createContext();

   export function PostsProvider({ children }) {
     const [page, setPage] = useState(0);
     const { posts, refetchPosts, loading, error } = useFetchPosts(page);
     const { handleCreatePost } = useCreatePost();
     const { handleUpdatePost } = useUpdatePost();
     const { handleDeletePost } = useDeletePost();

     const value = useMemo(() => ({
       posts,
       loading,
       error,
       createPost: handleCreatePost,
       updatePost: handleUpdatePost,
       deletePost: handleDeletePost,
       refetchPosts,
       page,
       goToPreviousPage: () => setPage(page - 1),
       goToNextPage: () => setPage(page + 1),
     }), [posts, loading, error, page]);

     return (
       <PostsContext.Provider value={value}>
         {children}
       </PostsContext.Provider>
     );
   }
   ```

### **Testing & Validation**:
- **Manual Testing**: Verified that all CRUD operations are working as expected with the API.
- **Edge Cases**: Tested edge cases such as attempting to update or delete non-existent posts, ensuring the application handles these scenarios gracefully.

### **Next Steps**:
- **Further Optimization**: Continue monitoring performance, especially as more data and application scales are added.
- **Additional Features**: Consider implementing more advanced features like error boundaries, or integrating pagination controls directly into the API requests.

### **Conclusion**:
This PR lays the foundation for managing posts in the application by establishing a robust connection to the API and optimizing state management. The application is ready to handle the core CRUD operations for posts, setting the stage for further enhancements and features.

### CODE REVIEW

**Review by trevorlana — APPROVED** [review]
seems good to me, and it would be nice to go over this as a group :)

**Review by CBlank-dev — APPROVED** [review]
